### PR TITLE
add obj true heading calc

### PR DESCRIPTION
**Pull request by adev4a**
# PR Details
This PR adds true heading calculation for an external object.
## Related Jira Key

[CDAR-252](https://usdot-carma.atlassian.net/browse/CDAR-252)

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

[CDAR-252]: https://usdot-carma.atlassian.net/browse/CDAR-252?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by willjohnsonk** [comment]
This is a new branch that has migrated from the original feature/cdar-252-external-object-to-sdsm branch which had git issues which were more easily avoided with a new instance. The original branch's PR can be found here: https://github.com/usdot-fhwa-stol/carma-platform/pull/2174

File: carma_cooperative_perception/include/carma_cooperative_perception/msg_conversion.hpp [hunk 1]
```diff
@@ -57,8 +57,9 @@ auto heading_to_enu_yaw(const units::angle::degree_t & heading) noexcept -> unit
 auto calc_relative_position(const carma_v2x_msgs::msg::Position3D & current_pose,
   const carma_v2x_msgs::msg::DetectedObjectData & detected_object_data) noexcept -> carma_v2x_msgs::msg::DetectedObjectData;
 
-auto transform_from_map_to_wgs84(const geometry_msgs::msg::PoseStamped source_pose, 
-  std::shared_ptr<lanelet::projection::LocalFrameProjector> map_projection) noexcept;
+auto transform_pose_from_map_to_wgs84(const geometry_msgs::msg::PoseStamped & source_pose, 
+  std::shared_ptr<lanelet::projection::LocalFrameProjector> map_projection) noexcept
```
[the remark this answers is not in the document]
Comment (willjohnsonk) [reply carma_cooperative_perception/include/carma_cooperative_perception/msg_conversion.hpp:61]: This is how I've seen it defined elsewhere- I think the projector needs it to be a shared_ptr for some reason? I'll double check that but I recall trying alternatives and it not wanting to cooperate.

**Review by adamlm — CHANGES_REQUESTED** [review]
Function name and return type changes look good. I was wrong about passing `std::shared_ptr` by value, apologies. If we are going to pass `std::shared_ptr` parameters, we should pass by reference-to-const (unless the underlying function needs to partake in the shared ownership, which I don't believe is the case here). Here's a [link](https://stackoverflow.com/questions/3310737/should-we-pass-a-shared-ptr-by-reference-or-by-value) to some more info.

Once that is updated, I think this will be good to go.

File: carma_cooperative_perception/include/carma_cooperative_perception/msg_conversion.hpp [hunk 2]
```diff
@@ -51,14 +51,11 @@ auto to_position_msg(const UtmCoordinate & position_utm) noexcept -> geometry_ms
...
 
-// TODO: Find a conversion from MAP pose to WGS84 heading
-// auto enu_orientation_to_wgs_heading(const geometry_msgs::msg::Quaternion & orientation) noexcept -> double;
-
 auto calc_relative_position(const carma_v2x_msgs::msg::Position3D & current_pose,
   const carma_v2x_msgs::msg::DetectedObjectData & detected_object_data) noexcept -> carma_v2x_msgs::msg::DetectedObjectData;
 
 auto transform_pose_from_map_to_wgs84(const geometry_msgs::msg::PoseStamped & source_pose, 
-  std::shared_ptr<lanelet::projection::LocalFrameProjector> map_projection) noexcept
+  const std::shared_ptr<lanelet::projection::LocalFrameProjector> map_projection) noexcept
```
Comment (adamlm) [inline carma_cooperative_perception/include/carma_cooperative_perception/msg_conversion.hpp:58]: This should be passed by reference, not value.

File: carma_cooperative_perception/src/msg_conversion.cpp [hunk 3]
```diff
@@ -173,7 +173,7 @@ auto calc_relative_position(const geometry_msgs::msg::PoseStamped & source_pose,
 }
 
 auto transform_pose_from_map_to_wgs84(const geometry_msgs::msg::PoseStamped & source_pose, 
-  std::shared_ptr<lanelet::projection::LocalFrameProjector> map_projection) noexcept
+  const std::shared_ptr<lanelet::projection::LocalFrameProjector> map_projection) noexcept
```
Comment (adamlm) [inline carma_cooperative_perception/src/msg_conversion.cpp:176]: This should be passed by reference, not value.

**Review by adamlm — CHANGES_REQUESTED** [review]

**Review by adamlm — APPROVED** [review]

**Pull request merged**